lib/plug/session/cookie.ex

```diff
@@ -68,43 +73,18 @@ defmodule Plug.Session.COOKIE do
 
   @impl true
   def init(opts) do
-    encryption_salt = opts[:encryption_salt]
-    signing_salt = check_signing_salt(opts)
-
-    iterations = Keyword.get(opts, :key_iterations, 1000)
-    length = Keyword.get(opts, :key_length, 32)
-    digest = Keyword.get(opts, :key_digest, :sha256)
-    log = Keyword.get(opts, :log, :debug)
-    secret_key_base = Keyword.get(opts, :secret_key_base)
-    key_opts = [iterations: iterations, length: length, digest: digest, cache: Plug.Keys]
-
-    serializer = check_serializer(opts[:serializer] || :external_term_format)
-
-    %{
-      encryption_salt: prederive(secret_key_base, encryption_salt, key_opts),
-      signing_salt: prederive(secret_key_base, signing_salt, key_opts),
-      key_opts: key_opts,
-      serializer: serializer,
-      log: log
-    }
+    build_opts(opts)
+    |> build_rotating_opts(opts[:rotating_options])
+    |> Map.delete(:secret_key_base)
   end
 
   @impl true
-  def get(conn, cookie, opts) do
-    %{key_opts: key_opts, signing_salt: signing_salt, log: log, serializer: serializer} = opts
+  def get(conn, raw_cookie, opts) do
+    opts = Map.put(opts, :secret_key_base, conn.secret_key_base)
```

Comment:
Instead of doing this, pass `conn.secret_key_base` to `read_raw_cookie` and we always fallback to reading `conn.secret_key_base` if none is set. WDYT?

Reply:
Thanks for the comment. Regarding  using `conn.secret_key_base` as fallback, it didn't feel great to me considering we're talking about rotating opts here. For example, let's say I want to add encryption, then intuitively I would provide the following config:

```exs
config :app, :session_options,
  signing_salt: "a/123",
  encryption_salt: "a/333",
  rotating_options: [
    [
      encryption_salt: nil,
      signing_salt: "a/123", # required
    ]
  ]
```

And it would work, but if next day someone changes `SECRET_KEY_BASE` then my rotation will quietly break.

I thought requiring `secret_key_base` on rotating opts explicitly helps avoid such accidents and makes the rotation process a bit safer.


And regarding `read_raw_cookie`, it's doing what the original `get` did, but I refactored it to use `opts.secret_key_base` instead of `conn.secret_key_base` such that it works with rotating opts as well. It then allowed me to reduce code required to support the rotation while also maintaining features like prederiving keys.